File: Assignment 5/Assignment5.java
import java.util.Scanner;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.Random;

import java.io.File;  
import java.io.FileNotFoundException;
import java.util.Scanner;

public class Assignment5{
	String fileName;
	ArrayList<String> inputData;
	HashMap<String, ArrayList<String>> grammar;

	public Assignment5(String fileName){
		this.fileName = fileName;
		this.inputData = readInputFromFile();
		this.grammar = formGrammarFromInputData();
	}

	public ArrayList<String> readInputFromFile(){
		ArrayList<String> inputData = new ArrayList<String>();
		try{
			File file = new File(this.fileName);
			Scanner myReader = new Scanner(file);
			while(myReader.hasNextLine()){
				String data = myReader.nextLine();
	        	inputData.add(data);
      		}
      		myReader.close();
		}
		catch(FileNotFoundException e) {
			System.out.println("An error occurred.");
      		e.printStackTrace();
		}
		return inputData;
	}

	public HashMap<String, ArrayList<String>> formGrammarFromInputData(){
		HashMap<String, ArrayList<String>> grammar = new HashMap<String, ArrayList<String>>();

		boolean isNextWordVariable = false;
		String currentVariableName = "";
		String production = "";
		for(int i = 0;i < this.inputData.size();i++){
			if(this.inputData.get(i).equals("{")){
				isNextWordVariable = true;
			}
			else if(isNextWordVariable){
				ArrayList<String> productions = new ArrayList<String>();
				grammar.put(this.inputData.get(i), productions);
				currentVariableName = inputData.get(i);
				isNextWordVariable = false;
			}
			else if(this.inputData.get(i).equals(";")){
				ArrayList<String> productions = grammar.get(currentVariableName);
				productions.add(production);
				production = "";
			}
			else if(!this.inputData.get(i).equals("}")){
				if(!production.equals("")) production += " ";
				production += this.inputData.get(i);
			}
		}

		return grammar;
	}

	public String randomSentence(){
		return randomSentence("<start>");
	}

	public String randomSentence(String variable){

		ArrayList<String> productions = this.grammar.get(variable);
		Random random = new Random();
		String production = productions.get(random.nextInt(productions.size()));
		String[] productionParts = production.split(" ");

		String sentence = "";
		for(int i = 0;i < productionParts.length;i++){
			if(this.grammar.containsKey(productionParts[i])){
				sentence += randomSentence(productionParts[i]);
			}
			else{
				sentence += productionParts[i];
			}
			if(i != productionParts.length - 1) sentence += " ";
		}
		return sentence;
	}

	public String toString(){
		String grammarInString = "\nGrammar Definition:\n\n";
		for(HashMap.Entry<String,ArrayList<String>> entry : this.grammar.entrySet()){
			String variable = entry.getKey();
			// System.out.println(variable);
			grammarInString += "\tVariable: " + variable + "\n";
			grammarInString += "\tProductions:\n";
			ArrayList<String> productions = entry.getValue();
			for(int i = 0;i < productions.size();i++){
				// System.out.println(productions.get(i));
				grammarInString += "\t\t" + productions.get(i) + "\n";
			}
			// System.out.println();
			grammarInString += "\n";
		}
		return grammarInString;
	}

	public static void main(String args[]){
		Scanner s = new Scanner(System.in);
		System.out.print("Enter the Grammar File Name: ");
		String fileName = s.next();

		Assignment5 rsg = new Assignment5(fileName);
		System.out.println(rsg.toString());
		String sentence1 = rsg.randomSentence();
		System.out.println("Sentence 1: " + sentence1);
		String sentence2 = rsg.randomSentence();
		System.out.println("Sentence 2: " + sentence2);
		String sentence3 = rsg.randomSentence();
		System.out.println("Sentence 3: " + sentence3);
	}
}

File: Assignment 5/Assignment5.cpp
#include <bits/stdc++.h>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <list>
#include <unordered_map>

using namespace std;

class Assignment5{
	public:
		string filename;
		list<string> inputData;
		unordered_map<string, list<string>> grammar;

		Assignment5(string filename){
			this->filename = filename;
			this->inputData = readInputFromFile();
			this->grammar = formGrammarFromInputData();
		}

		~Assignment5(){
			cout << "Destructing Class Variables\n";
		}

		list<string> readInputFromFile(){
			list<string> inputData;
			string data;
			ifstream myReader(this->filename);

			while (getline (myReader, data)) {
  				inputData.push_back(data);
			}

			myReader.close();

			// list<string>::iterator it;
    		// for (it = inputData.begin(); it != inputData.end(); ++it){
    		// 	cout << *it << "\n";
    		// }
			
			return inputData;
		}

		unordered_map<string, list<string>> formGrammarFromInputData(){
			unordered_map<string, list<string>> grammar;

			list<string> productions;
			bool isNextWordVariable = false;
			string currentVariableName = "";
			string production = "";

			for (auto itr: this->inputData){
				if(itr.compare("{") == 0){
					isNextWordVariable = true;
				}
				else if(isNextWordVariable){
					currentVariableName = itr;
					isNextWordVariable = false;
				}
				else if(itr.compare(";") == 0){
					productions.push_back(production);
					production = "";
				}
				else if(itr.compare("}") != 0){
					if(production.compare("") != 0) production += " ";
					production += itr;
				}
				else{
					list<string> productionsToAdd;
					for(auto i: productions){
						productionsToAdd.push_back(i);
					}

					grammar[currentVariableName] = productionsToAdd;
					productions.clear();
				}
			}

			return grammar;
		}

		string randomSentence(){
			return randomSentence("<start>");
		}

		string randomSentence(string variable){
			list<string> productions = this->grammar.at(variable);
			
			int value = rand() % productions.size();
			for(auto itr: productions){
				if(value == 0){
					string sentence = "";
					stringstream ss(itr);
					string word;
					while(ss >> word){
						if(this->grammar.find(word) != this->grammar.end()){
							sentence += randomSentence(word);
						}
						else{
							sentence += word;
							sentence += " ";
						}
					}
					return sentence;
				}
				value--;
			}
			return "";
		}

		friend ostream &operator<<(ostream &output, const Assignment5 &asg) { 
        	output << "\nGrammar Definition:\n\n";  
        	for(auto itr: asg.grammar){
        		string variable = itr.first;
        		list<string> productions = itr.second;
        		output << "\tVariable: " << variable << "\n";
        		output << "\tProductions:\n";
        		for(auto it: productions){
        			output << "\t\t" << it << "\n";
        		}
        		output << "\n";
        	}
        	return output;         
      	}

};

int main(){
	srand(time(0));
	string filename;
	cout << "Enter the Grammar File Name: ";
	cin >> filename;
	Assignment5 asg(filename);
	cout << asg;
	cout << "Sentence 1: ";
	cout << asg.randomSentence();
	cout << "\nSentence 2: ";
	cout << asg.randomSentence();
	cout << "\nSentence 3: ";
	cout << asg.randomSentence() << "\n";
}